hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -881,6 +881,196 @@ def draw_boolean(
         pass
 
 
+IRTreeNodeType: TypeAlias = Union["IRTreeNode", "IRTreeLeaf"]
+
+
+class IRTreeLeaf:
+    def __init__(
+        self,
+        *,
+        ir_type: IRTypeName,
+        value: IRType,
+        kwargs: IRKWargsType,
+        was_forced: bool,
+    ):
+        self.ir_type = ir_type
+        self.value = value
+        self.kwargs = kwargs
+        self.was_forced = was_forced
+
+    def _repr_pretty_(self, p, cycle):
+        assert cycle is False
+        p.text(f"{self.ir_type} {self.value} {self.kwargs}")
+
+    def copy(self, *, with_value: Optional[IRType] = None):
+        value = self.value if with_value is None else with_value
+        return IRTreeLeaf(
+            ir_type=self.ir_type,
+            value=value,
+            kwargs=self.kwargs,
+            was_forced=self.was_forced,
+        )
+
+    @property
+    def trivial(self):
+        """
+        A leaf is trivial if it cannot be simplified any further. This does not
+        mean that modifying a trivial leaf can't produce simpler test cases when
+        viewing the tree as a whole. Just that when viewing this leaf in
+        isolation, this is the simplest the leaf can get.
+        """
```

Comment:
Replacement for `Example.trivial` / `Blocks.trivial`. I think I'll actually remove this before merge, as I don't use it for `Float` (should I? we didn't previously), but I expect to use this general approach in future shrink pass conversions.